### PR TITLE
[node] cleaner definitions

### LINKED ISSUE #26908: [node] use extends keyof to document events
similar to how puppeteer documents there code - https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/puppeteer/index.d.ts#L1194

for example: in the `Worker` class
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/80d50f06cf6f6e075774c13f5fec8012a31529cb/types/node/index.d.ts#L1329-L1375

can we rewrittin as:
```ts
export interface WorkerEvents {
    disconnect: () => void;
    error: (error: Error) => void;
    exit: (code: number, signal: string) => void;
    listening: (address: Address) => void;
    message: (message: any, handle: net.Socket | net.Server) => void;  // the handle is a net.Socket or net.Server object, or undefined.
    online: () => void;
}

export class Worker {
    emit(event: string | symbol, ...args: any[]): boolean;
    emit(event: "disconnect"): boolean;
    emit(event: "error", error: Error): boolean;
    emit(event: "exit", code: number, signal: string): boolean;
    emit(event: "listening", address: Address): boolean;
    emit(event: "message", message: any, handle: net.Socket | net.Server): boolean;
    emit(event: "online"): boolean;

    addListener<K extends keyof WorkerEvents>(event: K, listener: WorkerEvents[K]): this;
    addListener(event: string, listener: (...args: any[]) => void): this;

    on<K extends keyof WorkerEvents>(event: K, listener: WorkerEvents[K]): this;
    on(event: string, listener: (...args: any[]) => void): this;

    once<K extends keyof WorkerEvents>(event: K, listener: WorkerEvents[K]): this;
    once(event: string, listener: (...args: any[]) => void): this;

    prependListener<K extends keyof WorkerEvents>(event: K, listener: WorkerEvents[K]): this;
    prependListener(event: string, listener: (...args: any[]) => void): this;

    prependOnceListener<K extends keyof WorkerEvents>(event: K, listener: WorkerEvents[K]): this;
    prependOnceListener(event: string, listener: (...args: any[]) => void): this;
}
```

does this make sense?

### PR DESCRIPTION
**Pull request by Meir017**
closes #26908

### CODE REVIEW

**Comment by Meir017** [comment]
should I also modify the `v0` ... `v9` directories?

**Comment by BrunoScheufler** [comment]
This PR is massive but I really like this approach!

**Review by BrunoScheufler — APPROVED** [review]

**Comment by Meir017** [comment]
I have no idea why the build failed, it doesn't seem related to the changes I made

https://travis-ci.org/DefinitelyTyped/DefinitelyTyped/builds/397803382#L3560
https://github.com/Meir017/DefinitelyTyped/blob/master/types/node/index.d.ts#L189

All of the errors seem to occurr only on `typescript@2.0` and not on `typescript@2.1`

**Comment by Meir017** [comment]
it shouldn't be merged yet

**Comment by Meir017** [comment]
@BrunoScheufler done

**Comment by Flarna** [comment]
On default CI verifies that typedefinitions work with all typescript versions since 2.0. Your change makes use of  `keyof` which was introduces with typescript 2.1 ==> test fails with 2.0.

You can specify the minimum supported typescript version by adding a comment like `// TypeScript Version: 2.1` as last element in header (see node/v8 typings).
But be aware that CI verifies also that all packages depending on your changed package require the same typescript version or a newer on. So it's likely the CI will complain again.

**Comment by Meir017** [comment]
I had another idea but it might be overkill

add something like:
```ts
export class EventEmitter<TEvents> {
    addListener(event: string | symbol, listener: (...args: any[]) => void): this;
    addListener<K extends keyof TEvents>(event: K, listener: TEvents[K]): this;
    on(event: string | symbol, listener: (...args: any[]) => void): this;
    on<K extends keyof TEvents>(event: K, listener: TEvents[K]): this;
    once(event: string | symbol, listener: (...args: any[]) => void): this;
    once<K extends keyof TEvents>(event: K, listener: TEvents[K]): this;
    removeListener(event: string | symbol, listener: (...args: any[]) => void): this;
    removeListener<K extends keyof TEvents>(event: K, listener: TEvents[K]): this;
    off(event: string | symbol, listener: (...args: any[]) => void): this;
    off<K extends keyof TEvents>(event: K, listener: TEvents[K]): this;
    removeAllListeners(event?: string | symbol): this;
    removeAllListeners<K extends keyof TEvents>(event?: K): this;
    setMaxListeners(n: number): this;
    getMaxListeners(): number;
    listeners(event: string | symbol): Function[];
    listeners<K extends keyof TEvents>(event: K): Function[];
    rawListeners(event: string | symbol): Function[];
    rawListeners<K extends keyof TEvents>(event: K): Function[];
    emit(event: string | symbol, ...args: any[]): boolean;
    listenerCount(type: string | symbol): number;
    listenerCount<K extends keyof TEvents>(type: K): number;
    prependListener(event: string | symbol, listener: (...args: any[]) => void): this;
    prependListenerr<K extends keyof TEvents>(event: K, listener: TEvents[K]): this;
    prependOnceListener(event: string | symbol, listener: (...args: any[]) => void): this;
    prependOnceListenerr<K extends keyof TEvents>(event: K, listener: TEvents[K]): this;
    eventNames(): Array<string | symbol>;
}
```

and then you can just extend it instead of adding multiple methods...

**Comment by Flarna** [comment]
Regarding minimum typescript version: There is #26410 ongoing which increases it to 2.3 or 2.4. So maybe just wait.

File: types/node/v4/index.d.ts [hunk 1]
```diff
@@ -1609,48 +1617,20 @@ declare module "net" {
...
+        addListener(event: string, listener: (...args: any[]) => void): this;
+        addListener<K extends keyof ServerEvents>(event: K, listener: ServerEvents[K]): this;
 
-        once(event: string, listener: Function): this;
-        once(event: "close", listener: () => void): this;
-        once(event: "connection", listener: (socket: Socket) => void): this;
-        once(event: "error", listener: (err: Error) => void): this;
-        once(event: "listening", listener: () => void): this;
-
-        prependListener(event: string, listener: Function): this;
```
Comment (lierdakil) [inline types/node/v4/index.d.ts]: Is this an intended omission, or an oversight? The old version has `prependListener` and `prependOnceListener` here, while the new one does not.

Comment (Meir017) [reply types/node/v4/index.d.ts]: Those are new in v6
Looks like they were added by mistake in the first place

**Comment by mhegazy** [comment]
To update the version of node, you will need to touch all its dependencies, that is likely to be half of the packages on DefinitelyTyped. It might be easier to just update the minimum version of TS we are testing against in this case.

**Comment by Meir017** [comment]
Waiting for #26410

**Pull request closed without merging**